### PR TITLE
Integrate Unidoc with TypelevelSitePlugin

### PR DESCRIPTION
**Pull request by armanbilge**
Closes https://github.com/typelevel/sbt-typelevel/issues/108.

- adds a `TypelevelUnidocPlugin` that kicks-in if the `ScalaUnidocPlugin` is enabled on the docs project
- API docs are copied into the Laika site, but only if this is a release tag (not a snapshot nor a pre-release tag)
- deploying the site now keeps existing files, so API docs persist even if they are not updated
- `homepage` and `apiURL` settings are populated automatically

One bit of trickiness is that even with this plugin enabled, your API docs URL will 404 until you do a tag release. This makes sense—in theory, you'd have to go back to the tag for the last stable release and deploy from there.

A temporary workaround is to either manually upload docs to the appropriately directory or set `tlSiteApiUri` to e.g. a javadoc.io url. Or just tag a release :)

### CODE REVIEW

File: site/src/main/scala/org/typelevel/sbt/TypelevelUnidocPlugin.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          homepage.value.map(_.toURI.resolve(api).toURL)
+      }
+    }
+  )
+
+  override def projectSettings = Seq(
+    laikaIncludeAPI := {
+      // include the API docs, only if this is not a snapshot nor a pre-release
+      // so long as tlSiteKeepFiles is true, it won't delete existing API docs
+      !isSnapshot.value && V.unapply(version.value).exists(!_.isPrerelease)
```
Comment (armanbilge) [inline site/src/main/scala/org/typelevel/sbt/TypelevelUnidocPlugin.scala:50]: I'd like to make this configurable in the future. Feeling a bit overwhelmed with settings—should API docs be updated for snapshot pushes? pre-releases? How about the `VERSION` variable injected into mdocs, should it be the last stable version, or the latest tagged version including pre-releases?

File: site/src/main/scala/org/typelevel/sbt/TypelevelSitePlugin.scala [hunk 2]
```diff
@@ -128,15 +148,16 @@ object TypelevelSitePlugin extends AutoPlugin {
             .toAbsolutePath
             .relativize(((Laika / target).value / "site").toPath)
             .toString,
-          "publish_branch" -> "gh-pages"
+          "keep_files" -> tlSiteKeepFiles.value.toString
         ),
         name = Some("Publish site"),
         cond = {
-          val predicate = tlSitePublishBranch
-            .value // Either publish from branch or on tags, not both
```
Comment (armanbilge) [inline site/src/main/scala/org/typelevel/sbt/TypelevelSitePlugin.scala:136]: When I first wrote this, I was afraid about race conditions when publishing the same commit twice, for a branch and for a tag. But I think it's fine.

File: build.sbt [hunk 3]
```diff
@@ -146,9 +146,13 @@ lazy val site = project
...
-  .dependsOn(kernel, githubActions, noPublish)
+  .dependsOn(kernel, github, githubActions, noPublish)
 
 lazy val docs = project
   .in(file("mdocs"))
   .enablePlugins(TypelevelSitePlugin)
-  .settings(laikaConfig ~= { _.withRawContent })
+  .enablePlugins(ScalaUnidocPlugin)
+  .settings(
+    tlSiteApiUri := Some(uri("api/org/typelevel/sbt/")),
```
Comment (armanbilge) [inline build.sbt:156]: Overriding this so we are teleported to the right package in the docs. Seems hard to configure automatically, but it's only a nice-to-have, not critical.

**Comment by rossabaker** [comment]
This deserves a more thoughtful review, but how does all this play with the idea of just pointing at javadoc.io?  We're publishing our own scaladoc to the site _and_ pointing at that?

**Comment by armanbilge** [comment]
Pointing at javadoc.io works fine for single-module projects, but not currently for multi-module projects that want all their scaladocs nicely condensed. I did wonder if we should publish an additional "unidoc" artifact to sonatype, precisely for this purpose.

What this plugin does, if enabled, is publish scaladocs to gh-pages and update all the API urls to point there instead of the javadoc.io default in the other PR. Which reminds me, I need to make sure the URLs account for versioning...

**Comment by armanbilge** [comment]
> I did wonder if we should publish an additional "unidoc" artifact to sonatype, precisely for this purpose.

Expanding on this: this would provide stable unidocs for _every_ published version, rather than a e.g. `3.x` that is updated every release. The downside imo is that we wouldn't be hosting docs at our own domain 🙃

**Comment by rossabaker** [comment]
javadoc.io may go away someday, but doc jars published to Maven Central should be around as long as Scala is around.  That Unidoc artifact is an interesting idea.

A lot of the bloat in the http4s repo is from all the years of publishing docs on each commit.  A good solution that got away from that would be nifty.

**Comment by armanbilge** [comment]
Cool, I'll looking into the unidoc artifact idea a bit. It could simplify things nicely.

> A lot of the bloat in the http4s repo is from all the years of publishing docs on each commit. A good solution that got away from that would be nifty.

Scaladocs in particular, or docs in general? The gh-pages deploy action we are using here only makes a commit if something changes (although, running mdoc does always change some non-determinstic outputs).

**Comment by rossabaker** [comment]
Scaladocs in particular.  Those generate big diffs from commit to commit with all the indexes and what not.

**Comment by armanbilge** [comment]
Btw publishing a unidoc artifact is very feasible. All you need is an project with the unidoc plugin enabled and this setting:

```scala
Compile / packageDoc / mappings := (ScalaUnidoc / packageDoc / mappings).value
```

I can pursue this strategy instead.

**Comment by armanbilge** [comment]
As it stands, this PR is largely superseded by https://github.com/typelevel/sbt-typelevel/pull/119 and https://github.com/typelevel/sbt-typelevel/pull/124.

The changes unique to this PR the direct integration between unidoc and laika, i.e.:
1. Copying the API docs directly into the laika site and deploying them together.
2. Automatically adding an API doc button to the site that points to the right place.

I'd definitely like to find a way to solve (2), but I need to think more about how to do it. Is (1) still valuable in a world where unidocs are viewable from javadoc.io? It depends how much a project cares about hosting docs at their own domain.

**Comment by armanbilge** [comment]
I have an idea to fix (2) in https://github.com/typelevel/sbt-typelevel/pull/119#discussion_r801864152. I'm going to punt (1) since folks seem to like the published unidoc concept.

With that, this PR has run its course.

**Pull request closed without merging**